deepspeech.go

```diff
@@ -63,9 +63,51 @@ type sliceHeader struct {
 // buffer     A 16-bit, mono raw audio signal at the appropriate sample rate.
 // bufferSize The number of samples in the audio signal.
 // sampleRate The sample-rate of the audio signal.
-// TODO Make sure the C string is cleaned properly
 func (m *Model) SpeechToText(buffer []int16, bufferSize, sampleRate uint) string {
-	return C.GoString(C.STT(m.w, (*C.short)(unsafe.Pointer((*sliceHeader)(unsafe.Pointer(&buffer)).Data)), C.uint(bufferSize), C.uint(sampleRate)))
+	str := C.STT(m.w, (*C.short)(unsafe.Pointer((*sliceHeader)(unsafe.Pointer(&buffer)).Data)), C.uint(bufferSize), C.uint(sampleRate));
+	retval := C.GoString(str)
+	C.FreeString(str)
+	return retval
+}
+
+type MetadataItem C.struct_MetadataItem
+
+func (mi *MetadataItem) Character() string {
+	return C.GoString(C.GetCharacter((*C.MetadataItem)(unsafe.Pointer(mi))))
+}
+
+func (mi* MetadataItem) Timestep() int {
+	return int(C.GetTimestep((*C.MetadataItem)(unsafe.Pointer(mi))))
+}
+
+func (mi* MetadataItem) StartTime() float32 {
+	return float32(C.GetStartTime((*C.MetadataItem)(unsafe.Pointer(mi))))
+}
+
+// Metadata represents a DeepSpeech metadata output
+type Metadata struct {
```

Comment:
Why don't you use the same method as `MetadataItem` which means :

1) create a`Metadata` type that wraps C.struct_Metadata
2) add relevant getters and setters that implement the same logic as in `toGoStruct`

Reply:
Because I kind of struggled at first with `Metadata`, then played with `MetadataItem` and was not sure what would be the best solution, so I felt leaving the two and having your opinion on what was a better solution.

Reply:
I feel your pain, CGO can be a real pain in the ass sometimes.

In all honesty, I like your `MetadataItem` solution better and I think the same should be done for `Metadata` if possible.

Reply:
Right, I'll move `struct Metadata` to it!

Reply:
@asticode I'm having a look at that right now, how are we supposed to handle the case of `MetadataItem` when using `type Metadata struct C.struct_Metadata` ? It's going to have a type of `[]C.struct_MetadataItem` ?

Reply:
Ok, looking at the output of `cgo`, this might not be the biggest issue:
```
type Metadata struct {
        Items           *MetadataItem
        Items           int32
        Pad_cgo_0       [4]byte
        Probability     float64
}
```

This is what gets generated when adding `type Metadata C.struct_Metadata` and the matching call.

Reply:
Okay, turns out I have something working with that solution @asticode !